### ISSUE TITLE
Question: is there a standard way of providing time information in CoAP ACK messages?

### ISSUE THREAD

**Issue by StFS** [eclipse-californium/californium#2134]
The title kinda says it all.

Is there some standard way of piggybacking time information with the response messages that CoAP sends to ACK CON requests?

We have been discussing ways to update and keep device clocks in sync without using expensive protocols such as NTP. Essentially it would be enough for us to just get a "correct" time as a response whenever we do CON requests as part of the ACK response.

I'm wondering whether there is some standard way of doing that and then, whether Californium supports that?

Or do I just need to define my own format for this data and put that into the payload of the ACK response (if that's even possible)?

Any suggestion on how to implement this would be nice.

p.s. I did find this but it seems to be a bit of overkill for what we need, and doesn't look like its been standardized at all:
http://kresttechnology.com/krest-academic-projects/krest-mtech-projects/ECE/M%20Tech-ECE%20EMBEDDED%202016-17/BASE%20PAPER/21.%20A%20Time%20Synchronization%20Technique%20for%20CoAP-based%20Home%20Automation%20Systems.pdf

**Comment by boaks** [comment]
I'm not aware of a standard way. I considered to ask the [IETF core mailing list](https://mailarchive.ietf.org/arch/browse/core/) for that last  years, but I missed it. Maybe it's now worth to get a feedback from there.

I was faced that calendar time synchronization task as well implementing the CoAP-S3-proxy demo.
My conclusion was to use a custom option, see [TimeOption](https://github.com/boaks/californium/blob/add_s3_proxy/demo-apps/cf-cloud-demo-server/src/main/java/org/eclipse/californium/cloud/option/TimeOption.java) (this requires Californium 3.8 with the new Option API).  

For now, I support that option resource specific, see [Devices](https://github.com/boaks/californium/blob/add_s3_proxy/demo-apps/cf-cloud-demo-server/src/main/java/org/eclipse/californium/cloud/resources/Devices.java#L289-L299) how to use it. 

The device is considered to add that `TimeOption` to any request (+ 8 bytes). The benefit will be, that the time of the sensor values will then not depend on the transmission time. If the server receives such a `TimeOption`, it will place also a `TImeOption` in the Response, if the difference is above a threshold. The value 0 from the device indicates, that the device has no proper time and the server is intended to send such a `TimeOption`, if the server supports that feature. 

I considered  to support also other time domains, e.g. "last change", "create", "valid" (recently updated without changing the value). 

For now, I don't consider the RTT too much. For my requirements, a resolution of 5-10s is well enough. Using cellular transmission it takes anyway up to a couple of seconds until the transmission actually happens. Better resolution would then require in many cases an extra exchange, when the "registration and route" is refreshed.
 
Any considerations/requirements from your side?

**Comment by StFS** [comment]
Thanks for the response.

Your implementation is a bit too elaborate for our needs, basically, all we need is to get the server time as part of the ACK for any CON request from a device so that the device can adjust its clock accordingly.

We're using the Hono CoAP adapter which uses californium and we're thinking about adding a feature flag to it so that if it is set (for example `HONO_COAP_ENABLETIMESTAMPRESPONSES` or something), all CON ACK messages would include the server timestamp.

I've hacked together a little experiment based on your code and I see that we should be able to use something very similar and I might end up getting that to a state where we will do a PR to the Hono CoAP adapter.

I just have one question about the option number you're using `0xff3c`. Is there anything specific behind that option number or just something you picked? I know there are rules for how they're defined depending on whether they are considered optional/mandatory/safe-to-forward etc, but in this case of the time option, would it be "safe" for us to just use the same option number you're using or should we put more effort into selecting it? If so, are there any specific considerations we should be addressing?

**Comment by boaks** [comment]
[IANA - CoAP Option Numbers](https://www.iana.org/assignments/core-parameters/core-parameters.xhtml#option-numbers)

It's in the "65000-65535 | Reserved for Experimental Use". I used "elective", though my device initially sends that to the server. If the server supports it well, if not, it gets ignored. I used "save" to have it forwarded.
And accidentally "nocachekey", because it doesn't define the resource. I will change that to "CacheKey" in order to by-pass a cache, because a cache doesn't make sense for a time option. I don't use proxies, but who knows.
If you want to send it only in the response, the "cachekey" is not considered (at least I think so ;-) ).
The messaging idea of hono breaks also proxy caches, though these caches are resource oriented. Beside of the OSCORE hype, proxies are not that often used.

Just one remark, because you use the term "ACK".
A empty ACK is empty. It has 4 bytes and ends with the MID. No options are allowed.
A piggybacked response ACK is a response, it may have options and payload.

Though the hono implementation was initially developed by me, I remember, I used both variants. The piggybacked response, if hono get's the response fast, and an ACK and separate response, if not.

So, you may send the option in a response (either ACK or CON), but not in a empty ACK.

**Comment by boaks** [comment]
Maybe it's good idea to agree and document ranges within the "65000-65535 | Reserved for Experimental Use" for Californium, Hono, and user custom-extensions.
Not sure, if leshan would also benfit from that. 

I will ask the other projects.

**Comment by boaks** [comment]
My current consideration:
(530 / 3 => 175)

- 65000 - 65174 (hono experiments)
- 65175 - 65349 (californium experiments)
- 65350 - 65535 (users experiments)

**Comment by StFS** [comment]
I have to apologize in advance for my ignorance when it comes to this.

I think I'm understanding what you're saying about the "empty" ACK vs piggybacked response ACK. I think that will be fine in my Hono implementation, if the client requests to get the server time (either with a request option or a request parameter) it will never get an "empty" ACK back, instead it will simply get the piggybacked response ACK.

However, I'm having a terrible time understanding the segmentation of the option numbers like you suggested above. At first I thought (from reading the RFC) that the option number was just a single byte and was very confused as to how you managed to get into the 65000 - 65535 range. But then I realized that I was probably only looking at the part of the option number that defines the properties of the option but there is probably a second byte there somewhere (that I am too tired now to confirm in the RFC 😉).

But I'm still struggling with this but let's see if I am getting this right.

So for a time option in the "hono experiments" range, I could use for example: `0xFE20` (which is 65056 in decimal)

The 2 byte binary representation of that is: `1111 1110 0010 0000`
or, more clearly represented in a "CoAP option number context": `1111 1110 001 000 0 0`
meaning that its:

- Elective (last 0 in the binary sequence)
- Safe-to-forward (second-to-last 0 in the binary)
- Is a CacheKey (the third-to-last to fifth-to-last values in the binary sequence not being all 1's)

The rest of the bits were flipped experimentally to put this into the 65000 - 65174 range.

So, am I getting this close to correct?

Note that I'm still not quite sure what the significance of it being a "cache-key" is exactly but as I said, I'm getting tired now so I'll just accept the fact that I don't quite understand it and move on for now :wink:.

**Comment by boaks** [comment]
[RFC 7252 3.1. Option Format](https://www.rfc-editor.org/rfc/rfc7252#section-3.1)

That specifies a delta + variable length encoding for the option number (the variable length is similar to something as UTF-8). It includes up to 16 bits. 

[RFC 7252 12.2 CoAP Option Numbers Registry](https://www.rfc-editor.org/rfc/rfc7252#section-12.2)

Describes the option number ranges.

>  a single byte

[RFC 7252 5.4.6 Option Numbers](https://www.rfc-editor.org/rfc/rfc7252#section-5.4.6)

"Figure 10: Option Number Mask (Least Significant Byte)"

Yes, a first look may end up in 1 byte, but it's two.

> So for a time option in the "hono experiments" range, I could use for example: 0xFE20 (which is 65056 in decimal)

hono seems to have removed the "already defined option", so I prefer to swap the ranges.

65000 - 65174 (californium experiments)
65175 - 65349 (hono experiments)
65350 - 65535 (users experiments)

> "hono experiments" 

As I wrote in a comment to your PR,  

> Just to mention:
> If it's mainly the same semantic, you may just use the number of Californium.

So, if encoding the milliseconds since 1970.01.01 is OK for you, I would got for the same number. It's not a that big difference, that a server may send that without being requested by the client. 

About the "features" of that option:
FMPOV it's "elective", it should not fail the request. 

The rest is more or less not relevant for hono, though AFAIK hono is not in line with REST, a proxy will anyway hardly work. A coap proxy is also not that frequently used. In a quite a lot of cases, CoAP traffic doesn't need such proxies and mainly uses common UDP/IP components as "border router".  Anyway, to align with it.

A proxy may cache GET responses in order to reuse that.
If so, other methods (DELETE, PUT, POST) invalid the cache, even for a slightly different resource returned in the "location option" (I also "misused" in hono for cmd-ctrl). 
In my understanding a proxy should forward the time option. The "no cache key" is then something special, it addresses proxies, which are not aware of the option number and therefore applies a "default processing". In my old variant I used that "no cache key". But considering, what should happen, I changed it. If it is a "cache key". With that, a GET request with a new "time" value will by pass the cache, which is the intention.  Any others, e.g. PUT or POST will anyway bypass the cache and may invalidate it.

I already adapted my numbers, but I didn't push that change. I will do that begin of next week.
I also started to ask at IANA, about to process of getting an official IANA number. If it's not to complicated, I will go for that.

**Comment by boaks** [comment]
Just to let you know:

Assigning  a IANA number requires effectively a "IETF review" (experience of the last 8 years).
The idea should be put in a RFC template and introduced to the CoRE WG.

For now, I would postpone to do that for the Time Option. Maybe I have more time later this year.

| Dec | Hex | Usage |
| - | - | :- |
| 65000 - 65174 | 0xFD8E0 - 0xFE96 | (californium experiments) |
| 65175 - 65349 | 0xFE97 - 0xFF45 | (hono experiments) |
| 65350 - 65535 | 0xFF46 - 0xFFFF | (users experiments) |

I use the number [0xfde8](https://github.com/boaks/californium/blob/add_cloud_demo_server/demo-apps/cf-cloud-demo-server/src/main/java/org/eclipse/californium/cloud/option/TimeOption.java#L49) for the time option.

One remark from the IANA expert:
65000-65535 is only intended for private experiments, not for public or interoperable usage or implementations.

**Comment by boaks** [comment]
If you really want to use different numbers, yes.

I would use the same number as Californium, as long as it is the system time in milliseconds since 1.1.1970.

**Comment by StFS** [comment]
Allright... I'll do that then.

**Comment by sophokles73** [comment]
> Note that I'm still not quite sure what the significance of it being a "cache-key" is exactly 

The option number used by @boaks marks the option as being part of a cache key (bits 3-5 are not all 1).

My understanding is that if a client sends a request including the TimeOption (value = 1) via a proxy, then the proxy may return a cached response that has been retrieved using a matching TimeOption value. However, the cached response's TimeOption in this case will most likely contain a value that is outdated because it contains the result of the computation that had been performed at the moment the original request had been processed. Or am I mistaken, @boaks?

If so, then FMPOV a CoAP server should mark all responses that contain the TimeOption as non-cachable by means of including the Max-Age = 0 option. Am I missing something @boaks?

Apart from this, I do not see a problem with re-using the option number as suggested by @boaks for your purposes, @StFS.

**Comment by boaks** [comment]
> My understanding

If a client uses a GET with TimeOption(value = 0), then a proxy may return the cached last response. 
In my idea, the client uses TimeOption(value = 0) only once after startup, and if this no GET, it's not affected by caching. Follow up request are then using TimeOption(value = <system time>), and that bypass the cache. Also all other requests (PUT or POST) will bypass the cache. Additionally, these bits describes only the default behavior, if the proxy doesn't support that option.

> If so, then FMPOV a CoAP server should mark all responses that contain the TimeOption as non-cachable by means of including the Max-Age = 0 option. Am I missing something @boaks?

That is also a good advice.

In sum: 
Therefore Klaus wrote, that in the experience of the last 8 years, always a expert review was required. 
And he recommended to put the stuff in RFC format to propose it to the core-wg.
Lets see, when I have some time left.

**Comment by boaks** [comment]
Could this issue be closed?

**Comment by StFS** [comment]
Absolutely. Thanks for all the great feedback and help!

Just a curiosity question: Is there any way for me to follow the discussion on how/whether to add a "time option" as one of the "official" CoAP options? Any Github issue or mailing list thread where that discussion is happening?

**Comment by boaks** [comment]
> Is there any way for me to follow the discussion on how/whether to add a "time option" as one of the "official" CoAP options? 

This would be mainly the [IETF core mailing list](https://mailarchive.ietf.org/arch/browse/core/).

There are also a couple of other source around coap, I tried to list some of them in the [wiki](https://github.com/eclipse-californium/californium/wiki/Links-to-CoAP-or-DTLS-1.2-research-information).
